File: LeetCode/Top-Interview-150/Binary-Search-Tree/minabsdiff.java
class TreeNode 
{
    int val;
    TreeNode left;
    TreeNode right;
    
    // Constructor for tree node
    TreeNode(int val) 
    { 
        this.val = val; 
    }
    
    // Constructor with left and right child nodes
    TreeNode(int val, TreeNode left, TreeNode right) 
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution 
{
    private TreeNode prev = null;
    private int minDiff = Integer.MAX_VALUE;

    public int getMinimumDifference(TreeNode root) 
    {
        inOrder(root);
        return minDiff;
    }

    private void inOrder(TreeNode node) 
    {
        if (node == null) return; // Base case for recursion
        
        // Traverse left subtree
        inOrder(node.left);

        // Process current node
        if (prev != null) 
        {
            minDiff = Math.min(minDiff, node.val - prev.val);
        }

        prev = node; // Update prev to current node for the next iteration

        // Traverse right subtree
        inOrder(node.right);
    }
}

// Driver code to test the Solution class
public class Main 
{
    public static void main(String[] args) 
    {
        TreeNode node1 = new TreeNode(1);
        TreeNode node3 = new TreeNode(3);
        TreeNode node2 = new TreeNode(2, node1, node3);
        TreeNode node6 = new TreeNode(6);
        TreeNode root = new TreeNode(4, node2, node6);

        Solution solution = new Solution();
        int minDiff = solution.getMinimumDifference(root);

        System.out.println("The minimum absolute difference in the BST is: " + minDiff);
    }
}

/*

Time Complexity:

- O(n)
- Each node is visited once during the InOrder Traversal.
- Constant amount of work is done when each node is visited (comparison, updating the min. diff and 
the current value to previous)

 */

File: LeetCode/Top-Interview-150/Binary-Search-Tree/minabsdiff.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.prev = None
        self.min_diff = float('inf')

        def inOrder(node):

            if not node:
                return
            
            # Left subtree
            inOrder(node.left)

            # Current node
            if self.prev is not None:

                self.min_diff = min(self.min_diff, node.val - self.prev)

            self.prev = node.val  # Update the current value to previous

            # Right subtree
            inOrder(node.right)

        inOrder(root)
        return int(self.min_diff)  

# Create the BST nodes
node1 = TreeNode(1)
node3 = TreeNode(3)
node2 = TreeNode(2, node1, node3)
node6 = TreeNode(6)
root = TreeNode(4, node2, node6)

# Instantiate Solution and find minimum difference
solution = Solution()
min_diff = solution.getMinimumDifference(root)

print(f"The minimum absolute difference in the BST is: {min_diff}")

'''

Time Complexity:

- O(n)
- Each node is visited once during the InOrder Traversal.
- Constant amount of work is done when each node is visited (comparison, updating the min. diff and 
the current value to previous)

'''